src/Features/Core/Portable/InvertConditional/AbstractInvertConditionalCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.InvertConditional
+{
+    internal abstract class AbstractInvertConditionalCodeRefactoringProvider<TConditionalExpressionSyntax> 
+        : CodeRefactoringProvider
+        where TConditionalExpressionSyntax : SyntaxNode
+    {
+        protected abstract bool ShouldOffer(TConditionalExpressionSyntax conditional, int position);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            if (span.Length > 0)
+            {
+                return;
+            }
+
+            var position = span.Start;
+            var conditional = await FindConditionalAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+            if (conditional == null ||
+                position < conditional.Span.Start ||
+                !ShouldOffer(conditional, position))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(
+                FeaturesResources.Invert_conditional,
+                c => InvertConditionalAsync(document, position, c)));
+        }
+
+        private static async Task<TConditionalExpressionSyntax> FindConditionalAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+
+            return token.Parent.FirstAncestorOrSelf<TConditionalExpressionSyntax>();
+        }
+
+        private async Task<Document> InvertConditionalAsync(
+            Document document, int position, CancellationToken cancellationToken)
+        {
+            var conditional = await FindConditionalAsync(document, position, cancellationToken).ConfigureAwait(false);
```

Comment:
It's an anti-pattern to capture nodes in the callbacks that actually do the refactoring.  If you do this, it means that the codefix can keep the *full* syntax-tree alive.  We try to avoid that so that the GC/caching system is free to release these trees.